src/framework/components/anim/controller.js

```diff
@@ -0,0 +1,560 @@
+                var offsetTime = activeState.getTimelineDuration() * transition.transitionOffset;
+                timeInState = offsetTime;
+                timeInStateBefore = offsetTime;
+            }
+            this.timeInState = timeInState;
+            this.timeInStateBefore = timeInStateBefore;
+        },
+
+        _transitionToState: function (newStateName) {
+            if (newStateName === this.activeStateName) {
+                return;
+            }
+
+            if (!this._getState(newStateName)) {
+                return;
+            }
+
+            var transition = this._findTransition(this.activeStateName, newStateName);
+            if (!transition) {
+                this.animEvaluator.removeClips();
+                transition = new AnimTransition(this, this.activeStateName, newStateName, 0, 0);
+            }
+            this._updateStateFromTransition(transition);
+        },
+
+        assignAnimation: function (stateName, animTrack) {
+            var state = this._getState(stateName);
+            if (!state) {
+                // #ifdef DEBUG
+                console.error('Linking animation asset to animation state that does not exist');
+                // #endif
+                return;
+            }
+
+            var animation = {
+                name: stateName + '.' + animTrack.name,
+                animTrack: animTrack,
+                weight: 1.0
+            };
+
+            // Currently the anim controller only supports single animations in a state
+            if (state.animations.length > 0) {
+                state.animations = [];
+                this.reset();
+            }
+            state.animations.push(animation);
+
+            if (!this.playing && this.activate && this.isPlayable()) {
+                this.play();
+
+            }
+        },
+
+        isPlayable: function () {
+            var playable = true;
+            var i;
+            for (i = 0; i < this.states.length; i++) {
+                if (!this.states[i].isPlayable()) {
+                    playable = false;
+                }
+            }
+            return playable;
+        },
+
+        play: function (stateName) {
+            if (stateName) {
+                this._transitionToState(stateName);
+            }
+            this.playing = true;
+        },
+
+        pause: function () {
+            this.playing = false;
+        },
+
+        reset: function () {
+            this.previousStateName = null;
+            this.activeStateName = ANIM_STATE_START;
+            this.playing = false;
+
+            this.currTransitionTime = 1.0;
+            this.totalTransitionTime = 1.0;
+            this.isTransitioning = false;
+
+            this.timeInState = 0;
+            this.timeInStateBefore = 0;
+
+            this.animEvaluator.removeClips();
+
+            this.parameters = Object.assign({}, this.initialParameters);
+        },
+
+        update: function (dt) {
+            if (this.playing) {
```

Comment:
Not a big deal but in general it's preferable around the codebase to early exit so that we reduce the amount of nested spacing. So this should be `if (!this.playing) return`.

Reply:
personally, I don't use early exits, I prefer this. It's a lot easier to profile code by adding markers around functions and similar.

Reply:
You can still add markers with early exits no?

Reply:
well if there's return in the middle of the function, then your endMarker at the end of function doesn't execute.

Reply:
Hmm, interesting point. We don't really have a coding standard about this AFAIK. I think @vkalpias has his personal style. I guess for me, I tend to use whatever structure works best for the function at hand.

Reply:
I'm not too bothered about this I'm just mentioning it for consistency. I've had comments from both Dave and Max in my previous PRs about this over the years so we mostly adopted that style. But again I don't mind

Reply:
We had places, where nested scopes were a bit out of hand, and for reducing nesting, as well as simplifying code reading (by reducing conditions of execution), we've followed early exits.

Reply:
I've moved this to use an early exit as it does make this function a fair bit more legible. I like the case for profiling though, it's something I wasn't considering.